Added search among testcases in html report

cakeinpanic opened the pull request:
We are using allure at work and suffer from lack of routing or at least search among test cases.
When we want to show someone specific allure report, we need to specify testcase and its parent, which is not very convenient

So i've added simple filtering on frontend, look how it works:
![preview](https://user-images.githubusercontent.com/588916/30683751-53e44638-9eb7-11e7-938f-30d1804b02bc.gif)

I also need to add some unit tests, but as soon as I decided not to store`searchQuery` at `settings`(cause it shouldn't be remembered on page refresh), I couldn't write units fast.
Maybe you can give me some advice about it?

Looking forward for merge to start using it)

just-boris commented on allure-generator/src/main/javascript/blocks/pane/styles.scss:
```diff
@@ -12,6 +12,10 @@
   &__title {
     padding: 0 15px 10px;
+    width:100%;
+    display: table;
```
Could we use `display: flex;` here? It would simplify the underlying code.

cakeinpanic replied:
of course! i couldn't find information about browser support for report, so did it in oldschool way)

just-boris commented on allure-generator/src/main/javascript/components/node-search/NodeSearchView.js:
```diff
@@ -0,0 +1,27 @@
+    initialize({onSearch, searchQuery}) {
+        this.onSearch = onSearch;
+
+        this.searchQuery = searchQuery;
```
You could pass a model here instead of `searchQuery/onSearch`.

@baev do you like it?

cakeinpanic replied:
by the way,  searchQuery is not needed to be passed, it left from the time I tried to re-rended this componentn on each state change

just-boris commented on allure-generator/src/main/javascript/components/tree/TreeView.js:
```diff
@@ -50,25 +40,16 @@ class TreeView extends View {
+        if (this.searchQuery) {
+            this.$('.node__title').each((i, node) => {
+                this.$(node).parent().addClass('node__expanded');
+            });
```
Something weird is happening here. 
@baev do we have expanded state stored in model?

cakeinpanic replied:
yes, when user clicks on a node, its state is stored, but i decided that better way is to force opening all nodes when search is in progress

cakeinpanic replied:
when search is finished, i restore previous state.
I already see bug here – if user toggles nodes by click while he is in search mode, this state is saved and then will be restored incorrectly.

cakeinpanic commented on allure-generator/src/main/javascript/components/node-search/NodeSearchView.js:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    onRender() {
+        this.$('input').val(this.searchQuery);
```
this should be gone, useless now

cakeinpanic commented:
Refactored using `settings` model, added unit test
We are really waiting for this feature, I'm interested in merging it)

@just-boris @baev @sseliverstov

just-boris commented on allure-generator/package.json:
```diff
@@ -7,7 +7,7 @@
     "start": "webpack-dev-server --inline --hot --env.development",
     "pretest": "eslint --ignore-path=.gitignore .",
-    "test": "jest",
+    "test": "jest --colors",
```
It breaks console output in Jenkins logs.
It is better not to commit such changes.

cakeinpanic replied:
oh wow, couldn't imagine that can break anythig. For local development is a musthave option, otherwise log is not readable(

there probably shuoul be added some gradlew flag or something like this, i'll research this

just-boris replied:
As Gradle is involved here, I have nothing to say. Usually I ran tests via `npm test` directly out of `allure-generator` folder.

just-boris commented on allure-generator/src/main/javascript/components/node-search/NodeSearchView.js:
```diff
@@ -0,0 +1,29 @@
+
+    close() {
+        this.settings.setSearchQuery('');
+    }
```
Is this used?

cakeinpanic replied:
not now, added just in case

just-boris commented on allure-generator/src/main/javascript/components/tree-view-container/TreeViewContainer.js:
```diff
@@ -0,0 +1,74 @@
+        }));
+
+        this.showChildView('search', new NodeSearchView({
+            settings: this.settings
```
Instead of settings, the `state` modal should be used. Then you will not need to introduce an extra `saveTemporary` method.

cakeinpanic replied:
I agree that extending `LocalStorageModel` is not the best way, but on the other side we store all settings(and searchQuery and filtering options are the same level settings) in one place. 
But i'll do as you say.

just-boris replied:
We are doing different separation of concerns here:
* settings is for all data, that should be saved as user preference
* state is a current state of UI (opened tree nodes, selected testcase, now is also search). It will not be saved in LocalStorage, but can be reflected in the URL, for example

cakeinpanic commented:
added and fixed display bug

cakeinpanic commented:
found and fixed display bug, which caused tree re-render on each node folding

sskorol commented:
looks cool 👍

cakeinpanic commented:
@baev looking forward for merge & release)

baev commented on allure-generator/src/main/javascript/data/tree/filter.js:
```diff
@@ -18,4 +17,24 @@ function byDuration(min, max) {
 
-export {byStatuses, byDuration};
\ No newline at end of file
+function byText(text) {
```
byName

cakeinpanic commented:
Okay, waiting!
Our testers literally cant work with allure reports because of this(

baev commented:
@cakeinpanic thanks!
